larpandoracontent/LArTwoDReco/LArCosmicRay/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,877 @@
+    const float predictedDownstreamOffset((predictedDownstreamPoint - downstreamPoint).GetMagnitude());
+
+    if (predictedDownstreamOffset > m_maxPredictedMergePointOffset)
+        return false;
+
+    const CartesianVector predictedUpstreamPoint(downstreamPoint + (downstreamDirection * separationDistance));
+    const float predictedUpstreamOffset((predictedUpstreamPoint - upstreamPoint).GetMagnitude());
+
+    if (predictedUpstreamOffset > m_maxPredictedMergePointOffset)
+        return false;
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::GetExtrapolatedCaloHits(const ClusterAssociation &clusterAssociation, const ClusterList *const pClusterList,
+    ClusterToCaloHitListMap &clusterToCaloHitListMap) const
+{
+    const CartesianVector &upstreamPoint(clusterAssociation.GetUpstreamMergePoint()), &downstreamPoint(clusterAssociation.GetDownstreamMergePoint());
+    const float minX(std::min(upstreamPoint.GetX(), downstreamPoint.GetX())), maxX(std::max(upstreamPoint.GetX(), downstreamPoint.GetX()));
+
+    for (const Cluster *const pCluster : *pClusterList)
+    {
+        const OrderedCaloHitList &orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+        for (const OrderedCaloHitList::value_type &mapEntry : orderedCaloHitList)
+        {
+            for (const CaloHit *const pCaloHit : *mapEntry.second)
+            {
+                const CartesianVector &hitPosition(pCaloHit->GetPositionVector());
+                if ((hitPosition.GetX() < minX) || (hitPosition.GetX() > maxX) || (hitPosition.GetZ() < upstreamPoint.GetZ()) || (hitPosition.GetZ() > downstreamPoint.GetZ()))
+                    continue;
+
+                const float distanceFromLine(clusterAssociation.GetConnectingLineDirection().GetCrossProduct(hitPosition - upstreamPoint).GetMagnitude());
+                if (distanceFromLine > m_distanceFromLine)
+                    continue;
+
+                clusterToCaloHitListMap[pCluster].push_back(pCaloHit);
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TrackInEMShowerAlgorithm::IsTrackContinuous(const ClusterAssociation &clusterAssociation, const ClusterToCaloHitListMap &clusterToCaloHitListMap) const
+{
+    // ATTN: Collect extrapolated calo hits and sort by projected distance from the upstreamMergePoint
+    CaloHitVector extrapolatedCaloHitVector;
+    for (const auto &entry : clusterToCaloHitListMap)
+        extrapolatedCaloHitVector.insert(extrapolatedCaloHitVector.begin(), entry.second.begin(), entry.second.end());
+
+    std::sort(extrapolatedCaloHitVector.begin(), extrapolatedCaloHitVector.end(), SortByDistanceAlongLine(clusterAssociation.GetUpstreamMergePoint(),
+        clusterAssociation.GetConnectingLineDirection()));
```

Comment:
Just have a quick think about whether it's likely you might get hits that have just about floating point identical sorting parameters; then it's not clear whether pairs of hits (which may come in in an arbitrary order) will always be ordered in the same manner. If the properties used to sort are within e.g. std::numeric_limits<float>::epsilon() of each other, it's often useful to have a tie-breaker available, such as the hit energy estimate (just pushes the problem to a higher order!). These are subtle things associated with working with unordered containers. Strictly, we ought to, across the board, work with vectors that hold their order (or are sorted, but will always be in the same way, as starting from defined initial order) and use the maps as look ups. We're doing OK here so long as we think about this carefully.